### PR TITLE
Implement a streaming API and tail in the fs command

### PR DESCRIPTION
**Pull request by dadgar**
This PR implements:
* Client.Stream API that allows streaming the contents of a file
* Updates the `fs` command to allow tailing files with `-f`, `-n` and `-c`

### CODE REVIEW

File: api/fs.go [hunk 1]
```diff
@@ -20,6 +28,18 @@ type AllocFileInfo struct {
 	ModTime  time.Time
 }
 
+// StreamFrame is used to frame data of a file when streaming
+type StreamFrame struct {
+	Offset    int64
+	Data      string
```
Comment (armon) [inline api/fs.go]: Change to `[]byte`

File: api/fs.go [hunk 2]
```diff
@@ -20,6 +28,18 @@ type AllocFileInfo struct {
...
+	File      string
+	FileEvent string
```
Comment (armon) [inline api/fs.go]: May as well `omitempty`

File: api/fs.go [hunk 3]
```diff
@@ -177,3 +197,72 @@ func (a *AllocFS) getErrorMsg(resp *http.Response) error {
 		return err
 	}
 }
+
+func (a *AllocFS) Stream(alloc *Allocation, path, origin string, offset int64,
```
Comment (armon) [inline api/fs.go]: Function comment

File: api/fs.go [hunk 4]
```diff
@@ -177,3 +197,72 @@ func (a *AllocFS) getErrorMsg(resp *http.Response) error {
...
+	c := http.Client{}
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create the output channel
+	frames := make(chan *StreamFrame, 10)
+
+	go func() {
```
Comment (armon) [inline api/fs.go]: Move to dedicated `streamDecode` function

File: client/allocdir/alloc_dir.go [hunk 5]
```diff
@@ -343,6 +360,38 @@ type ReadCloserWrapper struct {
...
+
+	return watcher.BlockUntilExists(t)
+}
+
+func (d *AllocDir) ChangeEvents(path string, curOffset int64, t *tomb.Tomb) (*watch.FileChanges, error) {
+	// Get the path relative to the alloc directory
+	p := filepath.Join(d.AllocDir, path)
+
+	var watcher watch.FileWatcher
+	if runtime.GOOS == "windows" {
```
Comment (armon) [inline client/allocdir/alloc_dir.go]: Refactor the watcher initialization to DRY

File: command/agent/fs_endpoint.go [hunk 6]
```diff
@@ -122,10 +166,398 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
...
 	return nil, nil
 }
+
+// StreamFrame is used to frame data of a file when streaming
+type StreamFrame struct {
+	// Offset is the offset the data was read from
+	Offset int64
+
+	// Data is the read data with Base64 byte encoding
+	Data string
```
Comment (armon) [inline command/agent/fs_endpoint.go]: Just use `[]byte`

File: command/agent/fs_endpoint.go [hunk 7]
```diff
@@ -122,10 +166,398 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
...
+	s.out.Close()
+	s.heartbeat.Stop()
+	s.flusher.Stop()
+}
+
+// Run starts a long lived goroutine that handles sending data as well as
+// heartbeating
+func (s *StreamFramer) Run() {
+	s.l.Lock()
+	s.running = true
```
Comment (armon) [inline command/agent/fs_endpoint.go]: Guard against double-Run

File: client/allocdir/alloc_dir.go (hunk 5 above)
Comment (sean-) [inline client/allocdir/alloc_dir.go]: With my nose in and eye on portability, can this be a helper function wrapped in a build tag?

Comment (sean-) [inline client/allocdir/alloc_dir.go]: Ditto above.

File: command/agent/fs_endpoint.go [hunk 8]
```diff
@@ -122,10 +166,398 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
...
+	s.l.Unlock()
+
+	go s.run()
+}
+
+// ExitCh returns a channel that will be closed when the run loop terminates.
+func (s *StreamFramer) ExitCh() <-chan struct{} {
+	return s.exitCh
+}
```
Comment (sean-) [inline command/agent/fs_endpoint.go]: I'm not sure why we don't do this more often, this is a nice trick.

File: command/agent/fs_endpoint.go [hunk 9]
```diff
@@ -122,10 +166,398 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
...
+
+// StreamFramer is used to buffer and send frames as well as heartbeat.
+type StreamFramer struct {
+	out       io.WriteCloser
+	enc       *codec.Encoder
+	frameSize int
+	heartbeat *time.Ticker
+	flusher   *time.Ticker
+	shutdown  chan struct{}
```
Comment (sean-) [inline command/agent/fs_endpoint.go]: Can you `go-rename` this to `shutdownCh` which is common elsewhere in the code?

File: command/fs.go [hunk 10]
```diff
@@ -212,19 +241,121 @@ nomad alloc-status %s`, allocID, allocID)
...
+	if !tail {
 		r, _, err := client.AllocFS().Cat(alloc, path, nil)
 		if err != nil {
 			f.Ui.Error(fmt.Sprintf("Error reading file: %s", err))
 			return 1
 		}
 		io.Copy(os.Stdout, r)
+	} else {
+		// Parse the offset
+		var offset int64 = int64(10) * bytesToLines
```
Comment (sean-) [inline command/fs.go]: Where did `10` come from?

File: command/fs.go [hunk 11]
```diff
@@ -1,17 +1,25 @@
...
 	"strings"
+	"syscall"
 	"time"
 
 	humanize "github.com/dustin/go-humanize"
 	"github.com/hashicorp/nomad/api"
 )
 
+const (
+	// bytesToLines is an estimation of how many bytes are in each log line.
```
Comment (sean-) [inline command/fs.go]: Can you add a small comment re: how this constant is used below?

File: command/fs.go [hunk 12]
```diff
@@ -212,19 +241,121 @@ nomad alloc-status %s`, allocID, allocID)
...
+// output is defaulted to 10 lines.
+func (f *FSCommand) followFile(client *api.Client, alloc *api.Allocation,
+	path string, offset int64) error {
+
+	cancel := make(chan struct{})
+	frames, _, err := client.AllocFS().Stream(alloc, path, api.OriginEnd, offset, cancel, nil)
+	if err != nil {
+		return err
+	}
+	signalCh := make(chan os.Signal, 3)
```
Comment (sean-) [inline command/fs.go]: `3` seems pretty arbitrary... is it the max number of expected inflight signals that are in queue to be delivered?  `3` -> constant?

File: command/fs.go [hunk 13]
```diff
@@ -212,19 +241,121 @@ nomad alloc-status %s`, allocID, allocID)
...
+
+		var err error
+		if follow {
+			err = f.followFile(client, alloc, path, offset)
+		} else {
+			// This offset needs to be relative from the front versus the follow
+			// is relative to the end
+			offset = file.Size - offset
+			r, _, err := client.AllocFS().ReadAt(alloc, path, offset, -1, nil)
+			if err != nil {
```
Comment (sean-) [inline command/fs.go]: ... if we're going to do `tail -F` here... [this is mandatory viewing for @diptanu and @dadgar](http://www.youtube.com/watch?v=vm1GJMp0QN4&t=40m46s) if you've never watched it before.

CC @jen20

File: command/fs.go (hunk 10 above)
Comment (dadgar) [inline command/fs.go]: Defaults to ten. Made a const

File: command/fs.go (hunk 13 above)
Comment (dadgar) [inline command/fs.go]: Yeah not doing -F.

File: api/fs.go [hunk 14]
```diff
@@ -177,3 +198,80 @@ func (a *AllocFS) getErrorMsg(resp *http.Response) error {
...
+		Scheme: "http",
+		Host:   node.HTTPAddr,
+		Path:   fmt.Sprintf("/v1/client/fs/stream/%s", alloc.ID),
+	}
+	v := url.Values{}
+	v.Set("path", path)
+	v.Set("origin", origin)
+	v.Set("offset", strconv.FormatInt(offset, 10))
+	u.RawQuery = v.Encode()
+	req := &http.Request{
```
Comment (diptanu) [inline api/fs.go]: Can we use the api client that we already have?

File: client/allocdir/alloc_dir.go [hunk 15]
```diff
@@ -333,6 +339,17 @@ func (d *AllocDir) ReadAt(path string, offset int64, limit int64) (io.ReadCloser
 	if _, err := f.Seek(offset, 0); err != nil {
 		return nil, fmt.Errorf("can't seek to offset %q: %v", offset, err)
 	}
+	return f, nil
+}
+
+// LimitReadAt returns a reader  for a file at the path relative to the alloc dir
+// which will read from a particular offset until EOF or limit is hit
+func (d *AllocDir) LimitReadAt(path string, offset int64, limit int64) (io.ReadCloser, error) {
```
Comment (diptanu) [inline client/allocdir/alloc_dir.go]: Doesn't add a lot of value.

File: command/agent/fs_endpoint.go [hunk 16]
```diff
@@ -122,10 +165,432 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
 		return nil, fmt.Errorf("file %q is a directory", path)
 	}
 
-	r, err := fs.ReadAt(path, int64(0), fileInfo.Size)
+	r, err := fs.ReadAt(path, int64(0))
```
Comment (diptanu) [inline command/agent/fs_endpoint.go]: Close the reader after we finish copying the contents.

File: command/agent/fs_endpoint.go [hunk 17]
```diff
@@ -122,10 +165,432 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
...
+// run is the internal run method. It exits if Destroy is called or an error
+// occurs, in which case the exit channel is closed.
+func (s *StreamFramer) run() {
+	// Store any error and mark it as not running
+	var err error
+	defer func() {
+		s.l.Lock()
+		s.err = err
+		s.out.Close()
+		close(s.exitCh)
```
Comment (diptanu) [inline command/agent/fs_endpoint.go]: Close the outbound channel as well.

**Comment by diptanu** [comment]
LGTM overall. Added a few comments, please resolve them before the PR is merged.

File: api/fs.go (hunk 14 above)
Comment (dadgar) [inline api/fs.go]: Not quite, requires the refactor I was telling you about.

File: command/agent/fs_endpoint.go [hunk 18]
```diff
@@ -122,10 +172,433 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
...
+		<-s.exitCh
+	}
+}
+
+// Run starts a long lived goroutine that handles sending data as well as
+// heartbeating
+func (s *StreamFramer) Run() {
+	s.l.Lock()
+	if s.running {
+		return
```
Comment (a86c6f7964) [inline command/agent/fs_endpoint.go]: Maybe I am wrong, but does this need to do Unlock before it returns?

**Pull request merged**